File: frontend/assets/js/analysis.js
// The report read backwards: every municipality ranked, then the buildings inside the
// official flood zones, each one a link to its own report.
//
// Everything shown here was computed in batch by backend/scripts/analysis_build.py. The
// page sorts and filters what the API hands it and never calculates a score of its own:
// two people opening the same ranking have to see the same list.

import { api } from "./api.js";
import { $, esc, setHtml } from "./util.js";

const PAGE = 50;

const state = {
  hazard: "overall",
  province: "",
  query: "",
  exposed: false,
  offset: 0,
  rows: [],
  total: 0,
  meta: {},
  hazards: [],
  provinces: [],
  isCount: false,
};

let app = null;
let searchTimer = null;
let inflight = null;

// The zones, worst first, as the official mapping names them.
const ZONE_LABEL = {
  zfp: "Preferential flow zone",
  t10: "Flood zone T=10 years",
  t50: "Flood zone T=50 years",
  t100: "Flood zone T=100 years",
  t500: "Flood zone T=500 years",
};

const USE_LABEL = {
  "1_residential": "Home",
  "2_agriculture": "Farm",
  "3_industrial": "Industrial",
  "4_1_office": "Office",
  "4_2_retail": "Retail",
  "4_3_publicServices": "Public service",
};

const levelClass = (level) => `lvl-${String(level || "no data").replace(/\s+/g, "-")}`;

const fmt = (n, digits = 0) =>
  n === null || n === undefined ? "—" : Number(n).toLocaleString("en", {
    minimumFractionDigits: digits, maximumFractionDigits: digits });

export function initAnalysis(application) {
  app = application;

  $("#anHazards").addEventListener("click", (e) => {
    const btn = e.target.closest("button[data-hazard]");
    if (!btn) return;
    setState({ hazard: btn.dataset.hazard, offset: 0 });
  });

  $("#anProvince").addEventListener("change", (e) =>
    setState({ province: e.target.value, offset: 0 }));

  $("#anQuery").addEventListener("input", (e) => {
    const value = e.target.value;
    clearTimeout(searchTimer);
    searchTimer = setTimeout(() => setState({ query: value, offset: 0 }), 250);
  });

  $("#anExposed").addEventListener("change", (e) =>
    setState({ exposed: e.target.checked, offset: 0 }));

  $("#anMore").addEventListener("click", () => load({ append: true }));

  $("#anTable").addEventListener("click", (e) => {
    const row = e.target.closest("[data-code]");
    if (row) openMunicipality(row.dataset.code);
  });

  const dialog = $("#anDialog");
  dialog.addEventListener("click", (e) => {
    if (e.target.closest("[data-close]")) closeDialog();
  });
  document.addEventListener("keydown", (e) => {
    if (e.key === "Escape" && !dialog.hidden) closeDialog();
  });
}

export function showAnalysis(params) {
  state.hazard = params.get("hazard") || "overall";
  state.province = params.get("province") || "";
  state.query = params.get("q") || "";
  state.exposed = params.get("exposed") === "1";
  state.offset = 0;
  $("#anQuery").value = state.query;
  $("#anExposed").checked = state.exposed;
  load({ append: false });
  const code = params.get("municipality");
  if (code) openMunicipality(code);
  else closeDialog({ silent: true });
}

function setState(patch) {
  Object.assign(state, patch);
  const qs = new URLSearchParams();
  if (state.hazard !== "overall") qs.set("hazard", state.hazard);
  if (state.province) qs.set("province", state.province);
  if (state.query) qs.set("q", state.query);
  if (state.exposed) qs.set("exposed", "1");
  const url = `/analisis${qs.toString() ? `?${qs}` : ""}`;
  history.replaceState({}, "", url);
  load({ append: false });
}

async function load({ append }) {
  if (inflight) inflight.abort();
  inflight = new AbortController();
  const offset = append ? state.offset + PAGE : 0;
  if (!append) setHtml($("#anTable"), `<p class="an-loading">Reading the ranking…</p>`);
  try {
    const data = await api.analysisRanking({
      hazard: state.hazard, province: state.province, q: state.query,
      exposed: state.exposed ? "true" : "", limit: PAGE, offset,
    }, inflight.signal);
    state.offset = offset;
    state.rows = append ? state.rows.concat(data.rows) : data.rows;
    state.total = data.total;
    state.meta = data.meta || {};
    state.hazards = data.hazards || [];
    state.provinces = data.provinces || [];
    state.isCount = !!data.is_count;
    render();
  } catch (err) {
    if (err && err.name === "AbortError") return;
    setHtml($("#anTable"), `<p class="an-empty">${esc(err.message)}</p>`);
    $("#anMore").hidden = true;
  } finally {
    inflight = null;
  }
}

function render() {
  renderHazards();
  renderProvinces();

  const meta = state.meta;
  $("#anMeta").textContent = state.total
    ? `${fmt(state.total)} municipalities · ${fmt(meta.with_exposure)} with their buildings `
      + `counted against the official flood maps · built ${meta.built || ""}`
    : "Nothing matches these filters.";

  const showsFlood = state.hazard === "flood" || state.isCount
    || state.rows.some((r) => r.exposure);
  const head = `
    <div class="an-row an-head">
      <span class="an-rank">#</span>
      <span class="an-name">Municipality</span>
      <span class="an-score">${esc(labelFor(state.hazard))}</span>
      <span class="an-fact">${showsFlood ? "Homes in a flood zone" : "What drives it"}</span>
    </div>`;

  // A count column is drawn against the biggest value on the page, not against 100:
  // the bar says "compared with the worst town here", which is what it is.
  const peak = state.isCount
    ? Math.max(1, ...state.rows.map((r) => r.homes_at_risk || 0)) : 100;

  const rows = state.rows.map((r) => {
    const value = state.isCount ? r.homes_at_risk : r.scores[state.hazard];
    const width = Math.max(2, Math.round(((value ?? 0) / peak) * 100));
    const shade = state.isCount ? scoreLevel(r.scores.flood) : scoreLevel(value);
    return `
      <button type="button" class="an-row" data-code="${esc(r.code)}">
        <span class="an-rank">${r.rank}</span>
        <span class="an-name">
          <strong>${esc(r.name)}</strong>
          <small>${esc(r.province)}${r.comarca ? ` · ${esc(r.comarca)}` : ""}</small>
        </span>
        <span class="an-score">
          <span class="an-bar ${levelClass(shade)}" style="width:${width}%"></span>
          <span class="an-num">${value === null || value === undefined ? "no data" : fmt(value, 0)}</span>
        </span>
        <span class="an-fact">${fact(r, showsFlood)}</span>
      </button>`;
  }).join("");

  setHtml($("#anTable"), head + (rows || `<p class="an-empty">Nothing matches these filters.</p>`));
  $("#anMore").hidden = state.rows.length >= state.total;

  setHtml($("#anNote"), note(meta));
}

function scoreLevel(score) {
  if (score === null || score === undefined) return "no data";
  if (score < 20) return "very low";
  if (score < 40) return "low";
  if (score < 60) return "moderate";
  if (score < 80) return "high";
  return "very high";
}

function labelFor(key) {
  const found = state.hazards.find((h) => h.key === key);
  return found ? found.label : key;
}

function fact(row, showsFlood) {
  if (showsFlood && row.exposure) {
    const e = row.exposure;
    return `<strong>${fmt(e.flooded_dwellings)}</strong> in ${fmt(e.flooded_buildings)} buildings`
      + `<small>${esc(ZONE_LABEL[e.worst_zone] || "")}</small>`;
  }
  const m = row.metrics || {};
  if (state.hazard === "fire_weather" && m.fire_weather_days !== undefined) {
    return `<strong>${fmt(m.fire_weather_days)}</strong> days a year of high fire danger`;
  }
  if (state.hazard === "flood_history" && m.flood_episodes !== undefined) {
    return `<strong>${fmt(m.flood_episodes)}</strong> flood episodes on record`;
  }
  if (state.hazard === "fire_history" && m.burnt_ha !== undefined) {
    return `<strong>${fmt(m.burnt_ha)}</strong> hectares burnt since 2011`;
  }
  if (state.hazard === "avalanche" && m.avalanche_zones !== undefined) {
    return `<strong>${fmt(m.avalanche_zones)}</strong> mapped avalanche zones`;
  }
  if (state.hazard === "heat" && m.heat_days_35 !== undefined) {
    return `<strong>${fmt(m.heat_days_35)}</strong> days a year above 35 °C`;
  }
  if (m.fire_weather_days !== undefined) {
    return `<strong>${fmt(m.fire_weather_days)}</strong> days a year of high fire danger`;
  }
  return "—";
}

function note(meta) {
  const missing = meta.not_covered;
  return `The ranking is built from what can be known for every town at once. `
    + `${fmt(meta.with_climate)} of ${fmt(meta.municipalities)} have the official `
    + `fire-danger grid; the flood column exists only for the ${fmt(meta.with_exposure)} `
    + `towns whose buildings have been crossed with the official maps, and heat only for `
    + `${fmt(meta.with_heat)}, because the climate record is rate-limited. Everything `
    + `else says "no data" instead of a low score.`
    + (missing ? ` The Basque provinces and Navarre are not here: ${esc(missing.reason)}.` : "")
    // The column is only as good as the mapping under it, and saying so is the
    // difference between a tool and a false reassurance.
    + ` And a low flood column is not a safe town: the official zones are studies of `
    + `mapped watercourses. Paiporta's urban area is classed T500 and Picanya has two `
    + `buildings mapped at all — both were flooded in October 2024.`;
}

function renderHazards() {
  const el = $("#anHazards");
  if (el.dataset.built === "1") {
    el.querySelectorAll("button").forEach((b) =>
      b.setAttribute("aria-selected", String(b.dataset.hazard === state.hazard)));
    return;
  }
  if (!state.hazards.length) return;
  setHtml(el, state.hazards.map((h) => `
    <button type="button" class="chip-btn" role="tab" data-hazard="${esc(h.key)}"
            aria-selected="${h.key === state.hazard}">${esc(h.label)}</button>`).join(""));
  el.dataset.built = "1";
}

function renderProvinces() {
  const el = $("#anProvince");
  if (el.dataset.built === "1" || !state.provinces.length) return;
  setHtml(el, `<option value="">All provinces</option>` + state.provinces.map((p) =>
    `<option value="${esc(p.code)}">${esc(p.name)}</option>`).join(""));
  el.value = state.province;
  el.dataset.built = "1";
}

// ------------------------------------------------------------------ one town

async function openMunicipality(code) {
  const dialog = $("#anDialog");
  dialog.hidden = false;
  document.body.classList.add("modal-open");
  $("#anTitle").textContent = "…";
  $("#anSub").textContent = "";
  setHtml($("#anDetail"), `<p class="an-loading">Reading the municipality…</p>`);
  try {
    const m = await api.analysisMunicipality(code);
    $("#anTitle").textContent = m.name;
    $("#anSub").textContent = [m.province, m.comarca].filter(Boolean).join(" · ");
    setHtml($("#anDetail"), detailHtml(m));
  } catch (err) {
    setHtml($("#anDetail"), `<p class="an-empty">${esc(err.message)}</p>`);
  }
}

function closeDialog({ silent = false } = {}) {
  const dialog = $("#anDialog");
  if (dialog.hidden) return;
  dialog.hidden = true;
  document.body.classList.remove("modal-open");
  if (!silent) $("#anQuery").blur();
}

function detailHtml(m) {
  const scores = Object.entries(m.scores)
    .filter(([, v]) => v !== null && v !== undefined)
    .sort((a, b) => b[1] - a[1]);

  const scoreList = `
    <div class="an-scores">
      ${scores.map(([k, v]) => `
        <div class="an-scorecell">
          <span class="an-scorekey">${esc(labelFor(k))}</span>
          <span class="an-bar ${levelClass(scoreLevel(v))}" style="width:${Math.max(2, v)}%"></span>
          <span class="an-num">${fmt(v)}</span>
        </div>`).join("")}
    </div>`;

  const exp = m.exposure && m.exposure.flood ? m.exposure : null;
  const exposure = exp ? `
    <section class="an-sec">
      <h3>The buildings that stand in an official flood zone</h3>
      <p class="an-secsub">
        ${fmt(exp.flood.buildings)} of ${fmt(exp.buildings.total)} buildings,
        ${fmt(exp.flood.dwellings)} dwellings, ${fmt(exp.flood.pre_1980)} of them in
        buildings raised before 1980. The cadastre's footprints crossed with the SNCZI
        flood zones the day this was built.
      </p>
      <div class="an-zones">
        ${exp.flood.by_zone.map((z) => `
          <div class="an-zone">
            <span class="an-zonename">${esc(z.label || ZONE_LABEL[z.zone] || z.zone)}</span>
            <span class="an-zonecount"><strong>${fmt(z.buildings)}</strong> buildings ·
              ${fmt(z.dwellings)} dwellings</span>
          </div>`).join("")}
      </div>
      <button type="button" class="chip-btn solid" id="anBuildings"
              data-code="${esc(m.code)}">See the buildings →</button>
      <div class="an-buildings" id="anBuildingList"></div>
    </section>` : `
    <section class="an-sec">
      <h3>The buildings have not been counted here yet</h3>
      <p class="an-secsub">
        Crossing this town's cadastral footprints with the official flood maps is a
        batch job: <code>analysis_build.py --only exposure --municipality ${esc(m.code)}</code>.
        Until it runs, this municipality has no flood column, and the ranking says so
        rather than assuming it is dry.
      </p>
    </section>`;

  const sources = `
    <section class="an-sec">
      <h3>Where every number comes from</h3>
      <ul class="an-sources">
        ${(m.indicators || []).map((i) => `
          <li>
            <span class="an-indname">${esc(i.label)}</span>
            <span class="an-indval">${fmt(i.value, i.unit === "ha" ? 1 : 0)} ${esc(i.unit)}</span>
            <small>${esc(i.provenance.source)} · ${esc(i.provenance.period)} ·
              ${esc(i.provenance.scale || "")}</small>
            <small class="an-method">${esc(i.provenance.method)}</small>
            ${i.context ? `<small class="an-ctx">${esc(i.context)}</small>` : ""}
          </li>`).join("")}
      </ul>
    </section>`;

  const own = m.centroid ? `
    <a class="chip-btn ghost" href="/report?address=${encodeURIComponent(m.centroid.join(", "))}"
       data-nav>Report for the centre of the town →</a>` : "";

  return scoreList + exposure + sources + `<div class="an-actions">${own}</div>`;
}

// The building list is loaded on demand: a town can have thousands of them.
document.addEventListener("click", async (e) => {
  const btn = e.target.closest("#anBuildings");
  if (!btn) return;
  const list = $("#anBuildingList");
  setHtml(list, `<p class="an-loading">Reading the buildings…</p>`);
  try {
    const data = await api.analysisBuildings(btn.dataset.code, { limit: 200 });
    setHtml(list, buildingsHtml(data));
    btn.hidden = true;
  } catch (err) {
    setHtml(list, `<p class="an-empty">${esc(err.message)}</p>`);
  }
});

function buildingsHtml(data) {
  if (!data.rows.length) return `<p class="an-empty">No exposed buildings.</p>`;
  return `
    <p class="an-secsub">The ${fmt(Math.min(data.rows.length, data.total))} worst of
      ${fmt(data.total)}. Each one opens its own report.</p>
    <div class="an-blist">
      ${data.rows.map((b) => `
        <a class="an-b" href="/report?address=${encodeURIComponent(`${b.lat}, ${b.lon}`)}" data-nav>
          <span class="an-bzone ${levelClass(scoreLevel(b.score))}">${esc(ZONE_LABEL[b.zone] || b.zone)}</span>
          <span class="an-bmain">
            <strong>${esc(USE_LABEL[b.use] || "Building")}${b.dwellings ? ` · ${fmt(b.dwellings)} dwellings` : ""}</strong>
            <small>${b.year ? `built ${b.year}` : "year unknown"}${b.area_m2 ? ` · ${fmt(b.area_m2)} m²` : ""}
              · ${esc(b.ref || "")}</small>
          </span>
          <span class="an-bgo">→</span>
        </a>`).join("")}
    </div>`;
}

File: backend/app/analysis.py
"""The report read backwards: every municipality of Spain, ranked, then its buildings.

The product answers "what threatens this home". This asks the same engine the opposite
question - "which homes does this threaten" - because a town hall, a comarca or an
emergency service cannot type eight thousand addresses, and the households who most
need the report are exactly the ones who will never go looking for it.

It is the same data and the same scales, at a different unit:

    municipality   the ranking. What can be known for all of Spain at once: the
                   official fire-danger grid at its centre, and, in Catalonia, the
                   flood episodes, wildfire perimeters and avalanche zones already on
                   disk for the reports.
    building       the drill-down. The cadastre's footprints for one municipality
                   crossed with the official flood zones that cover them, which is the
                   same question `providers/miteco.py` asks for one point, asked once
                   for every building in the town.

Nothing here is computed on the request path: `backend/scripts/analysis_build.py`
writes `data/analysis/` and this module only reads it. A ranking that recomputed
itself would be a different ranking every time it was opened, and the whole point is
a list somebody can act on and come back to.

What the ranking deliberately does NOT do:

  - It does not carry a heat score for the whole country. ERA5 is not available for
    eight thousand towns on a free tier, so heat is filled in only for the
    municipalities the build script was pointed at, and the rest say so rather than
    carry an invented number.
  - It does not rank people. The unit is a building, with what the cadastre publishes
    about the building: when it was built, what it is used for, how many dwellings it
    holds. Who lives in it is not in here and must not be put in here.
"""

from __future__ import annotations

import gzip
import json
from functools import lru_cache
from pathlib import Path

from . import config, scoring
from .indicators import Indicator, Provenance

DIR = config.ANALYSIS_DIR
MUNICIPALITIES = DIR / "municipalities.json.gz"
RANKING = DIR / "ranking.json.gz"
EXPOSURE_DIR = DIR / "exposure"
VERSION = 1

# The hazards the ranking can sort by. `overall` is the aggregate of the rest, with
# the same rule the report uses: the worst hazard rules and the others push up a little.
HAZARDS = ("overall", "people", "homes", "flood", "fire_weather", "flood_history",
           "fire_history", "avalanche", "heat")

# These two are not scores. They are the columns an institution actually plans around:
# how many dwellings, and how many residents, stand in an official flood zone, in
# absolute numbers. A small town can be a 92 and hold four hundred homes; a city can be
# a 70 and hold fifty thousand, and whoever has to reach them needs the second one
# first. `people` exists only where TALAIA has been asked (the `assets` build step).
COUNTS = {"homes": "homes_at_risk", "people": "people_at_risk"}

LABELS = {
    "overall": "Overall",
    "people": "People in a flood zone",
    "homes": "Homes in a flood zone",
    "flood": "Flooding (official zones, per building)",
    "fire_weather": "Wildfire weather",
    "flood_history": "Floods on record",
    "fire_history": "Wildfires on record",
    "avalanche": "Avalanches",
    "heat": "Extreme heat",
}

# --- Where each municipal number comes from ----------------------------------------

CATASTRO = dict(
    source="Dirección General del Catastro, INSPIRE download service",
    dataset="ES.SDGC.BU (buildings)",
    resolution="building footprint",
    url="https://www.catastro.hacienda.gob.es/INSPIRE/buildings/ES.SDGC.BU.atom.xml",
    scale="building",
    scale_kind="point",
)

CDS_FIRE_GRID = dict(
    source="Copernicus CDS · fire danger indicators for Europe (EURO-CORDEX)",
    dataset="sis-tourism-fire-danger-indicators",
    resolution="~11 km (EURO-CORDEX)",
    url="https://cds.climate.copernicus.eu/datasets/sis-tourism-fire-danger-indicators",
    scale="~11 km grid",
    scale_kind="grid",
)

TALAIA_AOI = dict(
    source="TALAIA · values at risk inside a polygon",
    dataset="talaia-v1-exposure",
    resolution="per asset; population from the 1 km INE census grid",
    url="https://talaia.up.railway.app",
    scale="the flooded part of the municipality",
    scale_kind="radius",
)

SNCZI_MUNICIPAL = dict(
    source="SNCZI (MITECO) flood zones crossed with the cadastre's building footprints",
    dataset="snczi-laminas × ES.SDGC.BU",
    resolution="official polygons at plot scale",
    url="https://gis.miteco.gob.es/geoserver/agua/ows",
    scale="municipality",
    scale_kind="municipality",
)


def _read(path):
    if not path.exists():
        return None
    with gzip.open(path, "rt", encoding="utf-8") as fh:
        return json.load(fh)


def _write(path, payload) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    with gzip.open(tmp, "wt", encoding="utf-8") as fh:
        json.dump(payload, fh, ensure_ascii=False)
    tmp.replace(path)


read = _read    # the build script reads and writes through the same door the API uses
write = _write


@lru_cache(maxsize=64)
def _cached(path_str: str, _mtime: float) -> dict | None:
    return _read(Path(path_str))


def _fresh(path) -> dict | None:
    """The file, parsed once and kept - but keyed on when it was last written.

    The ranking is rebuilt while the server is running. Caching it by path alone would
    serve the build from before the rebuild until someone restarted the process, and a
    ranking nobody can refresh is worse than no ranking.
    """
    try:
        mtime = path.stat().st_mtime
    except OSError:
        return None
    return _cached(str(path), mtime)


def ranking_data() -> dict | None:
    """The built ranking, or None when `analysis_build.py` has not run."""
    return _fresh(RANKING)


def universe() -> dict | None:
    return _fresh(MUNICIPALITIES)


def reload() -> None:
    """Forget what is cached in memory: the build script has written a new file."""
    _cached.cache_clear()


def available() -> bool:
    return ranking_data() is not None


def exposure(code: str) -> dict | None:
    """The building-by-building result for one municipality, if it has been built."""
    if not str(code).isdigit():
        return None
    return _fresh(EXPOSURE_DIR / f"{code}.json.gz")


# --- Queries -----------------------------------------------------------------------

def _score(row: dict, hazard: str) -> float:
    if hazard in COUNTS:
        return float(row.get(COUNTS[hazard]) or -1.0)
    value = (row.get("scores") or {}).get(hazard)
    return -1.0 if value is None else float(value)


def ranking(hazard: str = "overall", province: str | None = None, limit: int = 50,
            offset: int = 0, min_score: float | None = None,
            only_exposed: bool = False, query: str | None = None) -> dict:
    """The ranked municipalities, filtered.

    `hazard` picks the column to sort by, `only_exposed` keeps the ones whose buildings
    have been crossed with the flood zones, and `query` matches the name of the town.
    """
    data = ranking_data()
    if not data:
        return {"available": False, "total": 0, "rows": [], "hazards": list(HAZARDS)}

    hazard = hazard if hazard in HAZARDS else "overall"
    rows = data["rows"]
    if province:
        rows = [r for r in rows if r["province_code"] == province or r["province"] == province]
    if query:
        needle = _fold(query)
        rows = [r for r in rows if needle in _fold(r["name"])]
    if only_exposed:
        rows = [r for r in rows if r.get("exposure")]
    if hazard in COUNTS:
        # A count column lists the towns that HAVE the count. A town whose buildings
        # were never crossed with the flood maps does not hold zero people; it holds an
        # unknown number, and putting it in the list sorted as a zero would be a lie
        # told in alphabetical order.
        rows = [r for r in rows if r.get(COUNTS[hazard]) is not None]
    if min_score is not None:
        rows = [r for r in rows if _score(r, hazard) >= min_score]

    rows = sorted(rows, key=lambda r: (-_score(r, hazard), r["name"]))
    total = len(rows)
    page = rows[max(0, offset):max(0, offset) + max(1, min(limit, 500))]
    return {
        "available": True,
        "hazard": hazard,
        "label": LABELS.get(hazard, hazard),
        "hazards": [{"key": k, "label": LABELS[k]} for k in HAZARDS],
        "total": total,
        "offset": offset,
        # A count has no level: 47,089 homes is not "very high", it is 47,089 homes.
        "is_count": hazard in COUNTS,
        "rows": [{**r, "rank": offset + i + 1,
                  "level": None if hazard in COUNTS else scoring.level_for(_score(r, hazard))}
                 for i, r in enumerate(page)],
        "provinces": data.get("provinces", []),
        "meta": data.get("meta", {}),
    }


def _fold(text: str) -> str:
    """Lowercase and without accents, so "Alcala" finds "Alcalá"."""
    import unicodedata
    stripped = unicodedata.normalize("NFD", str(text))
    return "".join(c for c in stripped if unicodedata.category(c) != "Mn").lower()


def municipality(code: str) -> dict | None:
    """One municipality: its scores with their sources, and its buildings if built."""
    data = ranking_data()
    if not data:
        return None
    row = next((r for r in data["rows"] if r["code"] == str(code)), None)
    if not row:
        return None
    exp = exposure(code)
    return {
        **row,
        "indicators": [i.to_dict() for i in _indicators(row, exp)],
        "exposure": exp or row.get("exposure"),
        "meta": data.get("meta", {}),
    }


def _indicators(row: dict, exp: dict | None) -> list[Indicator]:
    """Every municipal number with where it comes from, the same contract as a report."""
    out: list[Indicator] = []
    metrics = row.get("metrics") or {}

    if metrics.get("fire_weather_days") is not None:
        out.append(Indicator(
            key="fire_weather_days",
            label="Days a year with a Fire Weather Index above 30",
            value=metrics["fire_weather_days"], unit="days/year",
            provenance=Provenance(period="1981-2005", confidence="medium",
                                  method="EURO-CORDEX multi-model mean, the grid cell "
                                         "nearest the centre of the municipality "
                                         "(within 25 km)",
                                  **CDS_FIRE_GRID),
            context=metrics.get("fire_weather_context"),
        ))

    if metrics.get("heat_days_35") is not None:
        out.append(Indicator(
            key="heat_days_35", label="Days a year above 35 °C",
            value=metrics["heat_days_35"], unit="days/year",
            provenance=Provenance(
                source="ERA5 / ERA5-Land (Copernicus C3S) via Open-Meteo Archive API",
                dataset="era5-land", period=f"{config.CLIMATOLOGY_START} to today",
                resolution="~9 km (ERA5-Land)",
                method="mean number of days with temperature_2m_max >= 35 °C at the "
                       "centre of the municipality, the same series a report reads",
                confidence="high", scale="~9 km grid", scale_kind="grid",
                url="https://open-meteo.com/en/docs/historical-weather-api"),
        ))

    if exp:
        total = exp["buildings"]["total"]
        out.append(Indicator(
            key="buildings_total", label="Buildings in the cadastre", value=total,
            unit="buildings",
            provenance=Provenance(period=exp["meta"]["cadastre_date"], confidence="high",
                                  method="every building of the municipality in the "
                                         "INSPIRE download, current at the date shown",
                                  **CATASTRO),
        ))
        out.append(Indicator(
            key="buildings_flooded",
            label="Buildings whose footprint centre falls in an official flood zone",
            value=exp["flood"]["buildings"], unit="buildings",
            provenance=Provenance(period="SNCZI, as published", confidence="medium",
                                  method="point-in-polygon of each footprint centre "
                                         "against the five SNCZI layers (preferential "
                                         "flow, T10, T50, T100, T500) over the "
                                         "municipality's bounding box",
                                  **SNCZI_MUNICIPAL),
            # From the counts, not from flood.share: that one is the cumulative share
            # of the zone that set the score, which is a different question.
            context=(f"{exp['flood']['buildings'] / total * 100:.1f} % of the town's "
                     f"buildings" if total else None),
        ))
        if exp["flood"]["dwellings"]:
            out.append(Indicator(
                key="dwellings_flooded", label="Dwellings in those buildings",
                value=exp["flood"]["dwellings"], unit="dwellings",
                provenance=Provenance(period=exp["meta"]["cadastre_date"],
                                      confidence="medium",
                                      method="numberOfDwellings of each exposed "
                                             "building, as the cadastre publishes it",
                                      **CATASTRO),
            ))

    assets = (exp or {}).get("assets") or {}
    if assets.get("available"):
        inside = assets["in_flood_zone"]
        out.append(Indicator(
            key="population_flooded", label="Residents of the flooded area",
            value=round(assets["population_resident"]), unit="people",
            provenance=Provenance(
                period=str(assets.get("generated_at") or "")[:10] or "as published",
                confidence="medium",
                method="INE census grid, area-weighted to the box around the buildings "
                       f"that stand in a flood zone ({assets['aoi']['km2']} km2). It "
                       "counts residents of that ground, not of the buildings",
                **TALAIA_AOI),
        ))
        out.append(Indicator(
            key="institutions_flooded",
            label="Schools, care homes and other places that hold people, inside a zone",
            value=inside["count"], unit="places",
            provenance=Provenance(
                period=str(assets.get("generated_at") or "")[:10] or "as published",
                confidence="medium",
                method="every asset TALAIA returned for the area, placed in a flood "
                       "zone by the same point-in-polygon test the buildings went "
                       "through",
                **TALAIA_AOI),
            context=(f"{inside['people']:.0f} people at capacity"
                     if inside["people"] else None),
        ))
        if assets.get("hazardous"):
            out.append(Indicator(
                key="hazardous_sites", label="Hazardous sites in the area",
                value=assets["hazardous"], unit="sites",
                provenance=Provenance(
                    period=str(assets.get("generated_at") or "")[:10] or "as published",
                    confidence="medium",
                    method="assets TALAIA flags as hazardous (fuel, chemicals, "
                           "industry holding dangerous substances)",
                    **TALAIA_AOI),
            ))
        if inside.get("value_eur"):
            out.append(Indicator(
                key="value_flooded", label="Replacement value inside a flood zone",
                value=round(inside["value_eur"]), unit="EUR",
                provenance=Provenance(
                    period=str(assets.get("generated_at") or "")[:10] or "as published",
                    confidence="low",
                    method="TALAIA's replacement valuation of the assets inside a "
                           "zone. A modelled cost from class defaults, not a survey "
                           "and not a market value",
                    **TALAIA_AOI),
            ))

    for key, label, period, method, source in _CATALAN_INDICATORS:
        value = metrics.get(key)
        if value is not None:
            out.append(Indicator(key=key, label=label, value=value, unit=source["unit"],
                                 provenance=Provenance(period=period, method=method,
                                                       confidence="high",
                                                       **source["provenance"])))
    return out


AGORA = dict(
    source="AGORA · University of Barcelona, flood episodes in Catalonia",
    dataset="agora-episodes", resolution="municipality",
    url="https://www.ub.edu/agora/", scale="municipality", scale_kind="municipality",
)
GENCAT_FIRE = dict(
    source="Government of Catalonia · wildfire perimeters",
    dataset="incendis-forestals", resolution="mapped perimeter",
    url="https://analisi.transparenciacatalunya.cat/", scale="municipality",
    scale_kind="municipality",
)
ICGC_AVALANCHE = dict(
    source="ICGC · avalanche zone map 1:25,000",
    dataset="allaus-zones", resolution="1:25,000",
    url="https://www.icgc.cat/", scale="municipality", scale_kind="municipality",
)

_CATALAN_INDICATORS = [
    ("flood_episodes", "Flood episodes that affected the municipality", "1902-2020",
     "count of AGORA episodes listing this municipality",
     {"unit": "episodes", "provenance": AGORA}),
    ("fires_count", "Wildfires recorded in the municipality", "2011 to today",
     "fires the Government of Catalonia lists for this municipality. Shown, not "
     "scored: counting fires ranks the towns with the most roadside grass first",
     {"unit": "fires", "provenance": GENCAT_FIRE}),
    ("burnt_ha", "Forest area burnt inside the municipality", "2011 to today",
     "sum of the forest hectares of those fires; this is what the wildfire-history "
     "score is built on",
     {"unit": "ha", "provenance": GENCAT_FIRE}),
    ("avalanche_zones", "Mapped avalanche zones in the municipality", "ICGC, as published",
     "avalanche zones whose centre falls inside the municipality. It says how much "
     "avalanche terrain was mapped here, never that a given house stands in one",
     {"unit": "zones", "provenance": ICGC_AVALANCHE}),
]


def buildings(code: str, zone: str | None = None, limit: int = 200,
              offset: int = 0) -> dict:
    """The exposed buildings of a municipality, worst zone first.

    Each one carries the coordinates the report takes as a query, which is the whole
    point of the drill-down: from a ranked town to a building to its own report.
    """
    exp = exposure(code)
    if not exp:
        return {"available": False, "total": 0, "rows": []}
    rows = exp.get("exposed") or []
    if zone:
        rows = [b for b in rows if b["zone"] == zone]
    total = len(rows)
    page = rows[max(0, offset):max(0, offset) + max(1, min(limit, 1000))]
    return {
        "available": True, "code": exp["code"], "name": exp["name"],
        "total": total, "offset": offset, "rows": page,
        "zones": exp["flood"]["by_zone"], "meta": exp["meta"],
    }
